Formula/postgresql@12.rb

```diff
@@ -81,22 +81,87 @@ def post_install
     return if ENV["CI"]
 
     (var/"log").mkpath
-    (var/"postgres").mkpath
-    unless File.exist? "#{var}/postgres/PG_VERSION"
-      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", "#{var}/postgres"
+    (var/name).mkpath
+    unless File.exist? "#{var}/#{name}/PG_VERSION"
+      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", "#{var}/#{name}"
     end
   end
 
+  # Previous versions of this formula used the same data dir as the regular
+  # postgresql formula. So we check whether the versioned data dir exists
+  # and has a PG_VERSION file, which should indicate that the versioned
+  # data dir is in use. Otherwise, returns the old data dir path.
+  def postgresql_datadir
+    if versioned_data_dir_exists?
+      "#{var}/#{name}"
+    else
+      "#{var}/postgres"
+    end
+  end
+
+  # Same as with the data dir - use old log file if the old data dir
+  # is version 12
+  def postgresql_log_path
+    if versioned_data_dir_exists?
+      "#{var}/log/#{name}"
+    else
+      "#{var}/log/postgres"
+    end
+  end
+
+  def versioned_data_dir_exists?
+    File.exist?("#{var}/#{name}/PG_VERSION")
+  end
+
+  def conflicts_with_postgresql_formula?
+    Formula["postgresql"].any_version_installed?
+  end
+
+  # Figure out what version of PostgreSQL the old data dir is
+  # using
+  def old_postgresql_datadir_version_12?
+    File.exist?("#{var}/postgres/PG_VERSION") &&
+      File.read("#{var}/postgres/PG_VERSION").chomp == "12"
+  end
+
   def caveats
-    <<~EOS
-      To migrate existing data from a previous major version of PostgreSQL run:
-        brew postgresql-upgrade-database
+    caveats = ""
+
+    # Check if we need to print a warning re: data dir
+    if old_postgresql_datadir_version_12?
+      caveats += if conflicts_with_postgresql_formula?
+        # Both PostgreSQL and PostgreSQL@12 are installed
+        <<~EOS
+          Previous versions of this formula used the same data directory as
+          the regular PostgreSQL formula. This causes a conflict if you
+          try to use both at the same time.
+
+        EOS
+      else
+        # Only PostgreSQL@12 is installed, not PostgreSQL
+        <<~EOS
+          Previous versions of this formula used the same data directory as
+          the regular PostgreSQL formula. This will cause a conflict if you
+          try to use both at the same time.
+
+          You can migrate to a versioned data directory by running this command:
 
+            mv -v "#{var}/postgres" "#{var}/#{name}"
+
+          (Make sure PostgreSQL is not running before executing this command)
+
+        EOS
+      end
+    end
+
+    caveats += <<~EOS
       This formula has created a default database cluster with:
-        initdb --locale=C -E UTF-8 #{var}/postgres
+        initdb --locale=C -E UTF-8 #{var}/#{name}
```

Comment:
Would be good to reuse one of the functions above.

Reply:
Addressed in 03dfafe1a4